hdr/sqlite_modern_cpp.h

```diff
@@ -16,515 +17,480 @@
-		}
-		error_occured = true;
-	}
+		template <typename Result>
+		typename std::enable_if<is_sqlite_value<Result>::value, void>::type operator>>(
+			Result& value) {
+			this->_extract_single_value([&value, this] {
+				get_col_from_db(*this, 0, value);
+			});
+		}
 
-	void throw_custom_error(const char* str) {
-		if(throw_exceptions) {
-			throw std::runtime_error(str);
+		template<typename... Types>
+		void operator>>(std::tuple<Types...>&& values) {
+			this->_extract_single_value([&values, this] {
+				tuple_iterate<std::tuple<Types...>>::iterate(values, *this);
+			});
 		}
-		error_occured = true;
-	}
 
-	template <typename Result>
-	typename std::enable_if<is_sqlite_value<Result>::value, void>::type operator>>(
-		Result& value) {
-		this->_extract_single_value([&value, this]{
-			get_col_from_db(*this,0, value);
-		});
-	}
+		template <typename Function>
+		typename std::enable_if<!is_sqlite_value<Function>::value, void>::type operator>>(
+			Function func) {
+			typedef utility::function_traits<Function> traits;
 
+			this->_extract([&func, this]() {
+				binder<traits::arity>::run(*this, func);
+			});
+		}
+	};
 
+	class database {
+	private:
+		std::shared_ptr<sqlite3> _db;
 
-	template<typename... Types>
-	void operator>>(std::tuple<Types...>&& values){
-		this->_extract_single_value([&values, this]{
-			tuple_iterate<std::tuple<Types...>>::iterate(values, *this);
-		});
-	}
+	public:
+		database(std::u16string const & db_name): _db(nullptr) {
+			sqlite3* tmp = nullptr;
+			auto ret = sqlite3_open16(db_name.data(), &tmp);
+			if(ret != SQLITE_OK) exceptions::throw_sqlite_error(ret);
+			_db = std::shared_ptr<sqlite3>(tmp, [=](sqlite3* ptr) { sqlite3_close_v2(ptr); ptr = nullptr; }); // close and null to be sure
+																											  //_db.reset(tmp, sqlite3_close); // alternative close. (faster?)
+		}
 
-	template <typename Function>
-	typename std::enable_if<!is_sqlite_value<Function>::value, void>::type operator>>(
-		Function func) {
-		typedef utility::function_traits<Function> traits;
+		database(std::string const & db_name):
+			database(std::u16string(db_name.begin(), db_name.end())) {}
 
-		this->_extract([&func, this]() {
-			binder<traits::arity>::run(*this, func);
-		});
-	}
-};
-
-class database {
-private:
-	sqlite3 * _db;
-	bool _connected;
-	bool _ownes_db;
-
-public:
-	database(std::u16string const & db_name):
-		_db(nullptr),
-		_connected(false),
-		_ownes_db(true) {
-		_connected = sqlite3_open16(db_name.data(), &_db) == SQLITE_OK;
-	}
+		database(std::shared_ptr<sqlite3> db):
+			_db(db) {}
+
+		//template<typename T>
+		//database_binder::chain_type operator<<(const T& sql) const {
+		//	return database_binder::chain_type(new database_binder(_db,std::string(sql)));
+		//}
+
+		database_binder::chain_type operator<<(const char* sql) {
+			return database_binder::chain_type(new database_binder(_db, std::string(sql)));
+		}
+
+		connection_type get_sqlite3_connection() const { return _db; }
```

Comment:
rename to  to `.connection()` . @Killili  does a setter makes sense ? (`.connection(shared_ptr<sqlite3*> db)`)